api/v1alpha1/externaldns_types.go

```diff
@@ -18,41 +18,412 @@ package v1alpha1
+	// requested hostname.
+	//
+	// This field accepts the following values:
+	//
+	//  "Exact": Explicitly match the full domain string
+	//   specified via the name field, including any subdomains
+	//   of the given name.
+	//
+	//  "Pattern": Match potential domains against
+	//  the provided regular expression pattern string.
+	//
+	// +unionDiscriminator
+	// +kubebuilder:validation:Required
+	// +required
+	MatchType DomainMatchType `json:"matchType"`
+
+	// Name is a string representing a single domain
+	// value.
+	//
+	// e.g. my-app.my-cluster-domain.com
+	//
+	// +optional
+	Name *string `json:"names,omitempty"`
+
+	// Pattern is a regular expression used to
+	// match a set of domains.
+	//
+	// +optional
+	Pattern *string `json:"pattern,omitempty"`
+}
+
+// +kubebuilder:validation:Enum=Exact;Pattern
+type DomainMatchType string
+
+const (
+	DomainMatchTypeExact DomainMatchType = "Exact"
+	DomainMatchTypeRegex DomainMatchType = "Pattern"
+)
+
+// +kubebuilder:validation:Enum=Include;Exclude
+type ExternalDNSFilterType string
+
+const (
+	FilterTypeInclude ExternalDNSFilterType = "Include"
+	FilterTypeExclude ExternalDNSFilterType = "Exclude"
+)
+
+// ExternalDNSProvider specifies configuration
+// options for the desired ExternalDNS DNS provider.
+type ExternalDNSProvider struct {
+	ExternalDNSProviderUnion `json:",inline"`
+
+	// Credentials is a reference to a secret containing
+	// the relevant credentials for the given provider, to be
+	// used by an ExternalDNS controller for creating, updating,
+	// and viewing DNS records.
+	//
+	// +kubebuilder:validation:Required
+	// +required
+	Credentials *corev1.LocalObjectReference `json:"credentials"`
+}
+
+// ExternalDNSProviderUnion describes optional fields for an ExternalDNS
+// provider that should be captured.
+// +union
+type ExternalDNSProviderUnion struct {
+	// Type describes which DNS provider
+	// ExternalDNS should publish records to.
+	// The following DNS providers are supported:
+	//
+	//  * AWS
+	//  * GCP
+	//  * Azure
+	//  * BlueCat
+	//  * Infoblox
+	//
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	// +required
+	Type ExternalDNSProviderType `json:"type"`
+}
+
+// +kubebuilder:validation:Enum=AWS;GCP;Azure;BlueCat;Infoblox
+type ExternalDNSProviderType string
+
+const (
+	ProviderTypeAWS      ExternalDNSProviderType = "AWS"
+	ProviderTypeGCP      ExternalDNSProviderType = "GCP"
+	ProviderTypeAzure    ExternalDNSProviderType = "Azure"
+	ProviderTypeBlueCat  ExternalDNSProviderType = "BlueCat"
+	ProviderTypeInfoblox ExternalDNSProviderType = "Infoblox"
+	// More providers will ultimately be added in the future.
+)
+
+// ExternalDNSSource describes which Source resource
+// the ExternalDNS should create DNS records for.
+type ExternalDNSSource struct {
+	ExternalDNSSourceUnion `json:",inline"`
+
+	// HostnameAnnotationPolicy specifies whether or not ExternalDNS
+	// should ignore the "external-dns.alpha.kubernetes.io/hostname"
```

Comment:
This seems to give considerable, unvalidated power to individual services or routes.

Reply:
Right, thats why we thought it would make sense to disable this functionality by default.

Reply:
Right. I beefed up the docs to include a warning against enabling this. Thoughts?

Reply:
> Right. I beefed up the docs to include a warning against enabling this. Thoughts?

Are you sure you even want this knob? Why not require a custom CR?

Reply:
> Why not require a custom CR?

A custom CR would be painful when creating external DNS records for LoadBalancer services.